notification-hubs-sdk/src/main/java/com/microsoft/windowsazure/messaging/notificationhubs/InstallationPutRequest.java

```diff
@@ -0,0 +1,192 @@
+package com.microsoft.windowsazure.messaging.notificationhubs;
+
+import android.os.Build;
+import android.util.Base64;
+
+import com.android.volley.NetworkResponse;
+import com.android.volley.Response;
+import com.android.volley.VolleyError;
+import com.android.volley.toolbox.HttpHeaderParser;
+import com.android.volley.toolbox.JsonObjectRequest;
+
+import org.json.JSONArray;
+import org.json.JSONException;
+import org.json.JSONObject;
+
+import java.io.UnsupportedEncodingException;
+import java.net.HttpURLConnection;
+import java.net.URLEncoder;
+import java.security.InvalidKeyException;
+import java.security.NoSuchAlgorithmException;
+import java.text.DateFormat;
+import java.text.SimpleDateFormat;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.Locale;
+import java.util.Map;
+
+import javax.crypto.Mac;
+import javax.crypto.spec.SecretKeySpec;
+
+/**
+ * Custom Android Volley Request to UPSERT an Installation records with the Azure Notification Hub backend.
+ *
+ * It was built specifically to work with API Version 2020-06.
+ */
+class InstallationPutRequest extends JsonObjectRequest {
+    private final static String API_VERSION = "2020-06";
+    private final static long TOKEN_EXPIRE_SECONDS = 5 * 60;
+    private final static DateFormat sIso8601Format = new SimpleDateFormat("yyyy-MM-dd'T'HH:mmZ", Locale.ENGLISH);
+
+    private final ConnectionString mConnectionString;
+
+    public InstallationPutRequest(ConnectionString connectionString, String hubName, Installation installation, Response.Listener<JSONObject> onSuccess, Response.ErrorListener onFailure){
+        super(
+                Method.PUT,
+                getInstallationUrl(connectionString.getEndpoint(), hubName, installation.getInstallationId()),
+                getBody(installation),
+                onSuccess,
+                onFailure);
+        mConnectionString = connectionString;
+    }
+
+    @Override
+    public Map<String, String> getHeaders() {
+        try {
+            Map<String,String> params = new HashMap<String, String>(){{
+                put("Content-Type", "application/json");
+                put("x-ms-version", API_VERSION);
+                put("Authorization", generateAuthToken(
+                        InstallationPutRequest.super.getUrl(),
+                        mConnectionString.getSharedAccessKeyName(),
+                        mConnectionString.getSharedAccessKey()));
+                put("User-Agent", getUserAgent());
+            }};
+            return params;
+        } catch (InvalidKeyException e) {

```

Comment:
Not sure why we need to catch this here if we're going to return null.  Would null break upstream?